apps/emqx_durable_storage/src/emqx_ds_snapshot_manager.erl

```diff
@@ -0,0 +1,428 @@
+    started_at :: _Time :: integer(),
+    dir :: file:filename(),
+    meta :: ra_snapshot:meta(),
+    state :: ra_state() | undefined,
+    reader_caller_proc :: {pid(), reference()} | undefined,
+    writer :: emqx_ds_storage_snapshot:writer() | undefined
+}).
+
+-type rs() :: #rs{}.
+-type ws() :: #ws{}.
+
+-type ext_rs() :: #{shard_id := shard_id()}.
+-type ext_ws() :: #{phase := machine_state, ws := ws()} | #{shard_id := shard_id()}.
+
+%% call/cast/info events
+-record(start_reader, {rs :: rs(), caller :: pid()}).
+-record(read_chunk, {size :: non_neg_integer()}).
+-record(begin_accept, {ws :: ws()}).
+-record(accept_chunk, {chunk :: binary()}).
+-record(complete_accept, {chunk :: binary()}).
+
+-export_type([ext_rs/0, ext_ws/0]).
+
+%%------------------------------------------------------------------------------
+%% API
+%%------------------------------------------------------------------------------
+
+-spec start_link(emqx_ds:db(), emqx_ds_replication_layer:shard_id()) -> {ok, pid()}.
+start_link(DB, ShardId) ->
+    gen_server:start_link(?via({DB, ShardId}), ?MODULE, #{db => DB, shard_id => ShardId}, []).
+
+-spec begin_read(_SnapshotDir :: file:filename(), _Context :: #{}) ->
+    {ok, ra_snapshot:meta(), ext_rs()} | {error, _Reason :: term()}.
+begin_read(Dir, _Context) ->
+    RS0 = #rs{
+        phase = machine_state,
+        started_at = erlang:monotonic_time(millisecond)
+    },
+    case ra_log_snapshot:recover(Dir) of
+        {ok, Meta, MachineState} ->
+            ShardId = shard_id(MachineState),
+            RS1 = RS0#rs{state = MachineState},
+            ok = gen_server:call(?via(ShardId), #start_reader{rs = RS1, caller = self()}, infinity),
+            ReadState = #{shard_id => ShardId},
+            {ok, Meta, ReadState};
+        Error ->
+            Error
+    end.
+
+-spec read_chunk(ext_rs(), _Size :: non_neg_integer(), _SnapshotDir :: file:filename()) ->
+    {ok, binary(), {next, ext_rs()} | last} | {error, _Reason :: term()}.
+read_chunk(#{shard_id := ShardId} = ExtRS, Size, _Dir) ->
+    case gen_server:call(?via(ShardId), #read_chunk{size = Size}, infinity) of
+        {ok, Chunk, next} ->
+            ?tp(dsrepl_read_chunk_next, #{}),
+            {ok, Chunk, {next, ExtRS}};
+        {ok, Chunk, last} ->
+            {ok, Chunk, last};
+        {error, Reason} ->
+            error(Reason)
+    end.
+
+-spec begin_accept(_SnapshotDir :: file:filename(), ra_snapshot:meta()) ->
+    {ok, ext_ws()}.
+begin_accept(Dir, Meta) ->
+    WS = #ws{
+        phase = machine_state,
+        started_at = erlang:monotonic_time(millisecond),
+        dir = Dir,
+        meta = Meta
+    },
+    ExtWS = #{phase => machine_state, ws => WS},
+    {ok, ExtWS}.
+
+-spec accept_chunk(binary(), ext_ws()) ->
+    {ok, ext_ws()} | {error, _Reason :: term()}.
+accept_chunk(Chunk, ExtWS = #{shard_id := ShardId}) ->
+    case gen_server:call(?via(ShardId), #accept_chunk{chunk = Chunk}, infinity) of
```

Comment:
There was no enforcing at this point.  I just moved most of the existing logic and ownership over the files to this module from the original one.  Would it be possible that `ra` would trigger more than one snapshot transfer on the same follower shard?

Reply:
> There was no enforcing at this point.

Yeah, I know, it's also a shortcoming of the original implementation.

> Would it be possible for `ra` to trigger more than one snapshot transfer on the same follower shard?

If you mean "initiate a snapshot writer", then I believe this is possible, this could (probably need to check 🫠) happen when leader changes while transfer is in progress.